File: sa-reminder-backend/src/main/java/com/workspace/sareminderbackend/controller/DepartmentController.java
package com.workspace.sareminderbackend.controller;

import com.mybatisflex.core.paginate.Page;
import com.workspace.sareminderbackend.annotation.AuthCheck;
import com.workspace.sareminderbackend.common.BaseResponse;
import com.workspace.sareminderbackend.common.DeleteRequest;
import com.workspace.sareminderbackend.common.ResultUtils;
import com.workspace.sareminderbackend.constant.UserConstant;
import com.workspace.sareminderbackend.exception.ErrorCode;
import com.workspace.sareminderbackend.exception.ThrowUtils;
import com.workspace.sareminderbackend.model.dto.department.DepartmentAddRequest;
import com.workspace.sareminderbackend.model.dto.department.DepartmentAssignUsersRequest;
import com.workspace.sareminderbackend.model.dto.department.DepartmentQueryRequest;
import com.workspace.sareminderbackend.model.dto.department.DepartmentTransferUsersRequest;
import com.workspace.sareminderbackend.model.dto.department.DepartmentUpdateRequest;
import com.workspace.sareminderbackend.model.entity.Department;
import com.workspace.sareminderbackend.model.entity.User;
import com.workspace.sareminderbackend.model.vo.DepartmentVO;
import com.workspace.sareminderbackend.model.vo.UserVO;
import com.workspace.sareminderbackend.service.DepartmentService;
import com.workspace.sareminderbackend.service.UserService;
import jakarta.annotation.Resource;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

@RestController
@RequestMapping("/department")
public class DepartmentController {

    @Resource
    private DepartmentService departmentService;

    @Resource
    private UserService userService;

    /**
     * 添加部门
     *
     * @param request
     * @return
     */
    @PostMapping("/add")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Long> addDepart(@RequestBody DepartmentAddRequest request) {
        ThrowUtils.throwIf(request == null, ErrorCode.PARAMS_ERROR);
        return ResultUtils.success(departmentService.addDepartment(request));
    }

    /**
     * 更新部门
     *
     * @param request
     * @return
     */
    @PostMapping("/update")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> updateDepart(@RequestBody DepartmentUpdateRequest request) {
        ThrowUtils.throwIf(request == null || request.getId() == null, ErrorCode.PARAMS_ERROR);
        return ResultUtils.success(departmentService.updateDepartment(request));
    }

    /**
     * 删除部门
     *
     * @param request
     * @return
     */
    @PostMapping("/delete")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> deleteDepart(@RequestBody DeleteRequest request) {
        ThrowUtils.throwIf(request == null || request.getId() == null || request.getId() <= 0, ErrorCode.PARAMS_ERROR);
        return ResultUtils.success(departmentService.deleteDepartment(request.getId()));
    }

    /**
     * 获取部门
     *
     * @param id
     * @return
     */
    @GetMapping("/get")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<DepartmentVO> getById(long id) {
        ThrowUtils.throwIf(id <= 0, ErrorCode.PARAMS_ERROR);
        Department department = departmentService.getValidDepartment(id);
        return ResultUtils.success(departmentService.getDepartmentVO(department));
    }

    /**
     * 获取部门列表
     *
     * @param request
     * @return
     */
    @PostMapping("/list/page/vo")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Page<DepartmentVO>> listPage(@RequestBody DepartmentQueryRequest request) {
        ThrowUtils.throwIf(request == null, ErrorCode.PARAMS_ERROR);
        long pageNum = request.getPageNum();
        long pageSize = request.getPageSize();
        Page<Department> page = departmentService.page(Page.of(pageNum, pageSize), departmentService.getQueryWrapper(request));
        Page<DepartmentVO> voPage = new Page<>(pageNum, pageSize, page.getTotalRow());
        voPage.setRecords(departmentService.getDepartmentVOList(page.getRecords()));
        return ResultUtils.success(voPage);
    }

    /**
     * 分配用户
     *
     * @param request
     * @return
     */
    @PostMapping("/assign/users")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> assignUsers(@RequestBody DepartmentAssignUsersRequest request) {
        ThrowUtils.throwIf(request == null || request.getDepartmentId() == null, ErrorCode.PARAMS_ERROR);
        return ResultUtils.success(departmentService.assignUsers(request));
    }

    /**
     * 转移用户
     *
     * @param request
     * @return
     */
    @PostMapping("/transfer/users")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<Boolean> transferUsers(@RequestBody DepartmentTransferUsersRequest request) {
        ThrowUtils.throwIf(request == null || request.getFromDepartmentId() == null || request.getToDepartmentId() == null,
                ErrorCode.PARAMS_ERROR);
        return ResultUtils.success(departmentService.transferUsers(request));
    }

    /**
     * 导入用户
     *
     * @param file
     * @param departmentId
     * @param defaultRole
     * @return
     */
    @PostMapping(value = "/import/users", consumes = {"multipart/form-data"})
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public BaseResponse<List<UserVO>> importUsers(@RequestPart("file") MultipartFile file,
                                                  @RequestParam("departmentId") Long departmentId,
                                                  @RequestParam(value = "defaultRole", required = false) String defaultRole) {
        ThrowUtils.throwIf(departmentId == null || departmentId <= 0, ErrorCode.PARAMS_ERROR);
        List<User> userList = departmentService.importUsersFromExcel(departmentId, defaultRole, file);
        return ResultUtils.success(userService.getUserVOList(userList));
    }
}

File: sa-reminder-backend/sa-reminder-frontend/src/api/departmentController.ts
// @ts-ignore
/* eslint-disable */
import request from '@/request'

/** 此处后端没有提供注释 POST /department/add */
export async function addDepart(body: API.DepartmentAddRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseLong>('/department/add', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /department/assign/users */
export async function assignUsers(
  body: API.DepartmentAssignUsersRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>('/department/assign/users', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /department/delete */
export async function deleteDepart(body: API.DeleteRequest, options?: { [key: string]: any }) {
  return request<API.BaseResponseBoolean>('/department/delete', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 GET /department/get */
export async function getById1(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getById1Params,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseDepartmentVO>('/department/get', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /department/import/users */
export async function importUsers(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.importUsersParams,
  body: {},
  file?: File,
  options?: { [key: string]: any }
) {
  const formData = new FormData()

  if (file) {
    formData.append('file', file)
  }

  Object.keys(body).forEach((ele) => {
    const item = (body as any)[ele]

    if (item !== undefined && item !== null) {
      if (typeof item === 'object' && !(item instanceof File)) {
        if (item instanceof Array) {
          item.forEach((f) => formData.append(ele, f || ''))
        } else {
          formData.append(ele, new Blob([JSON.stringify(item)], { type: 'application/json' }))
        }
      } else {
        formData.append(ele, item)
      }
    }
  })

  return request<API.BaseResponseListUserVO>('/department/import/users', {
    method: 'POST',
    params: {
      ...params,
    },
    data: formData,
    requestType: 'form',
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /department/list/page/vo */
export async function listPage1(
  body: API.DepartmentQueryRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponsePageDepartmentVO>('/department/list/page/vo', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /department/transfer/users */
export async function transferUsers(
  body: API.DepartmentTransferUsersRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>('/department/transfer/users', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** 此处后端没有提供注释 POST /department/update */
export async function updateDepart(
  body: API.DepartmentUpdateRequest,
  options?: { [key: string]: any }
) {
  return request<API.BaseResponseBoolean>('/department/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}
